### PR TITLE
[SPARK-25010][SQL] Rand/Randn should produce different values for each execution in streaming query

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Like Uuid in SPARK-24896, Rand and Randn expressions now produce the same results for each execution in streaming query. It doesn't make too much sense for streaming queries. We should make them produce different results as Uuid.

In this change, similar to Uuid, we assign new random seeds to Rand/Randn when returning optimized plan from `IncrementalExecution`.

Note: Different to Uuid, Rand/Randn can be created with initial seed. Because we replace this initial seed at `IncrementalExecution`, it doesn't use the initial seed anymore. For now it seems to me not a big issue for streaming query. But need to confirm with others. cc @zsxwing @cloud-fan 

## How was this patch tested?

Added test.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala [hunk 1]
```diff
@@ -75,14 +75,15 @@ class IncrementalExecution(
...
   override lazy val optimizedPlan: LogicalPlan = {
-    val random = new Random()
-
     sparkSession.sessionState.optimizer.execute(withCachedData) transformAllExpressions {
       case ts @ CurrentBatchTimestamp(timestamp, _, _) =>
         logInfo(s"Current batch timestamp = $timestamp")
         ts.toLiteral
       // SPARK-24896: Set the seed for random number generation in Uuid expressions.
-      case _: Uuid => Uuid(Some(random.nextLong()))
+      case _: Uuid => Uuid(Some(Utils.random.nextLong()))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala]: shall we create a trait `ExpressionWithRandomSeed` which has a `withNewSeed` method for these expressions?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala]: Sounds good. Let me update it accordingly.

**Comment by viirya** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala [hunk 2]
```diff
@@ -75,14 +74,11 @@ class IncrementalExecution(
    * with the desired literal
    */
   override lazy val optimizedPlan: LogicalPlan = {
-    val random = new Random()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala]: why do we remove this central `Random` for setting the seeds?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala]: I saw Rand/Randn use`Utils.random`, so changed to use it consistently.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 3]
```diff
@@ -854,6 +854,26 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging wi
     assert(uuids.distinct.size == 2)
   }
 
+  test("Rand/Randn in streaming query should not produce results in each execution") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: `produce results` -> `produce same results`

**Comment by cloud-fan** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: oops, fixed typo.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by zsxwing** [comment]
LGTM2

**Comment by HyukjinKwon** [comment]
LGTM3 😃

**Pull request closed without merging**